spacy/tests/doc/test_json_doc_conversion.py

```diff
@@ -57,18 +99,68 @@ def test_doc_to_json(doc):
     assert json_doc["ents"][0]["end"] == 3  # character offset!
     assert json_doc["ents"][0]["label"] == "ORG"
     assert not schemas.validate(schemas.DocJSONSchema, json_doc)
+    assert srsly.json_loads(srsly.json_dumps(json_doc)) == json_doc
 
 
 def test_doc_to_json_underscore(doc):
-    Doc.set_extension("json_test1", default=False)
-    Doc.set_extension("json_test2", default=False)
+    if not Doc.has_extension("json_test1"):
+        Doc.set_extension("json_test1", default=False)
+    if not Doc.has_extension("json_test2"):
+        Doc.set_extension("json_test2", default=False)
     doc._.json_test1 = "hello world"
     doc._.json_test2 = [1, 2, 3]
+
     json_doc = doc.to_json(underscore=["json_test1", "json_test2"])
     assert "_" in json_doc
     assert json_doc["_"]["json_test1"] == "hello world"
     assert json_doc["_"]["json_test2"] == [1, 2, 3]
     assert not schemas.validate(schemas.DocJSONSchema, json_doc)
+    assert srsly.json_loads(srsly.json_dumps(json_doc)) == json_doc
+
+
+def test_doc_to_json_with_token_span_attributes(doc):
+    if not Doc.has_extension("json_test1"):
+        Doc.set_extension("json_test1", default=False)
+    if not Doc.has_extension("json_test2"):
+        Doc.set_extension("json_test2", default=False)
+    if not Token.has_extension("token_test"):
+        Token.set_extension("token_test", default=False, force=True)
+    if not Span.has_extension("span_test"):
+        Span.set_extension("span_test", default=False, force=True)
+
+    doc._.json_test1 = "hello world"
+    doc._.json_test2 = [1, 2, 3]
+    doc[0:1]._.span_test = "span_attribute"
+    doc[0]._.token_test = 117
+    doc.spans["span_group"] = [doc[0:1]]
+    json_doc = doc.to_json(
+        underscore=["json_test1", "json_test2", "token_test", "span_test"]
+    )
+
+    assert "_" in json_doc
+    assert json_doc["_"]["json_test1"] == "hello world"
+    assert json_doc["_"]["json_test2"] == [1, 2, 3]
+    assert "user_data" in json_doc
+    assert json_doc["user_data"]["token_test"]["attribute_value"] == 117
+    assert json_doc["user_data"]["span_test"]["attribute_value"] == "span_attribute"
+    assert not schemas.validate(schemas.DocJSONSchema, json_doc)
+    assert srsly.json_loads(srsly.json_dumps(json_doc)) == json_doc
+
+
+def test_doc_to_json_with_token_span_attributes_missing(doc):
```

Comment:
Nitpick on the name: should it be "token_attribute_missing"? Because the span attribute is included, right?